rust/pyapi/src/services/jtag.rs

```diff
@@ -0,0 +1,185 @@
+use crate::extract_value;
+use crate::model::Model;
+use crate::{resolve_transaction, unpack_transaction_options};
+use origen::services::{jtag, Service};
+use pyo3::prelude::*;
+use pyo3::types::{PyAny, PyDict};
+
+#[pyclass]
+#[derive(Debug, Clone)]
+pub struct JTAG {
+    id: usize,
+}
+
+#[pymethods]
+impl JTAG {
+    #[new]
+    fn new() -> Self {
+        JTAG { id: 0 }
+    }
+
+    pub fn id(&self) -> PyResult<usize> {
+        Ok(self.id)
+    }
+
+    // This is called automatically by Origen whenever a service is registered, the object
+    // returned by this method is what get saved in the controller's services dict.
+    fn set_model(&mut self, name: &str, model: &Model) -> PyResult<Self> {
+        crate::dut::PyDUT::ensure_pins("dut")?;
+        let mut dut = origen::dut();
+        let mut services = origen::services();
+        let id = services.next_id();
+        let service = Service::JTAG(jtag::Service::new(
+            &dut, id, None, // default IR size
+            None, // tdi
+            None, // tdo
+            None, // toms
+            None, // tclk
+            None, // trstn
+        )?);
+        services.add_service(service);
+        model.materialize_mut(&mut dut)?.add_service(name, id)?;
+        self.id = id;
+        Ok(self.clone())
+    }
+
+    fn reset(&self) -> PyResult<()> {
+        let dut = origen::dut();
+        let services = origen::services();
+        let jtag = services.get_as_jtag(self.id)?;
+        jtag.reset(&dut)?;
+        Ok(())
+    }
+
+    #[args(write_opts = "**")]
+    fn write_register(&self, bits: &PyAny, write_opts: Option<&PyDict>) -> PyResult<()> {
```

Comment:
What does write_register do in the context of a JTAG driver?
Would this not be better called `shift_register`?
Generally I think the `write_register` / `read_register` API methods should be reserved for register-map-level drivers.

Reply:
This is most in-depth I've gotten with JTAG, and in much of the documentation I was looking at, the DR and IR were referred to as registers, even though they aren't registers in the usual sense. But, since I kept seeing them referred to as such, I added verify/write registers

Though, these will actually shift in/out the data and not leave the "register" in the same state. So probably a good update would be to either change the name as you suggest, or shift the assumed data in again so it maintains its state.

If we change to shift though, would we still want a shift_verify/write? Or just an option, as you had above? I'm fine with a name change, but would we want to keep the two-method approach consistent? I had planned to add a `capture` as well once I have that in order, but it could be a `capture_shift` or something instead?

Reply:
I think the main thing here is to change the name to something else. From the point of view of a user who doesn't really understand protocols all that well, if they see methods called `write/verify_register` then they might rightly assume that they can give this a DUT register object and expect it to generate a valid transation.
So we should consider these as special method names which are essentially a DUT register handler trait and which the JTAG driver obviously doesn't support.

I don't have any special preference on what other name(s) to use, nor really about consistency between physical protocol drivers. If anything, I probably prefer physical drivers to implement their own unique APIs because it emphasises that at this level Origen is not providing much abstraction and the user really needs to know what they are doing.

Reply:
Okeydoke. The functions to accept DUT register objects but its implied to be within the JTAG block. It was more for writing internal registers. I do already have a write/verify IR/DR so I guess that's sufficient then?

I don't actually use write/verify register anywhere. I added it for completeness, but going a bit too complete I guess. I just removed them unless you have another operation they should be replaced with. I think you can do everything you need to with just write/read IR/DR.

Reply:
Yep I'm fine with that, I think those IR/DR read/write methods are all we need here.